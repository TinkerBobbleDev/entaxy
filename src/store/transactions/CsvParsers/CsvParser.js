/* eslint-disable prefer-destructuring */
import { isNil } from 'lodash'
import Papa from 'papaparse'
import parse from 'date-fns/parse'

const DONT_IMPORT = 'Don\'t import'
export const DATE_FORMATS = {
  'mm/dd/yyyy': (string) => {
    const [, month, day, year] = (/(\d{1,2})\/(\d{1,2})\/(\d{4})/gi).exec(string)
    return [year, month, day]
  },
  yyyyymmdd: (string) => {
    const [, year, month, day] = (/(\d{4})(\d{2})(\d{2})/gi).exec(string)
    return [year, month, day]
  }
}
const TRANSACTION_FIELDS = {
  description1: {
    label: 'Description 1',
    regex: /Description|Name|Memo/gi
  },
  description2: {
    label: 'Description 2',
    regex: /Description|Name|Memo/gi
  },
  amount: {
    label: 'Amount',
    regex: /Amount/gi
  },
  income: {
    label: 'Income',
    regex: /Credit/gi
  },
  expense: {
    label: 'Expense',
    regex: /Debit/gi
  },
  createdAt: {
    label: 'Date',
    regex: /Date/gi
  }
}

// The base class for all CSV parsers
export default class CsvParser {
  constructor() {
    this._csvData = []
    this._csvHeader = []
    this._columnsCount = 0
    this._firstRowIndex = 0
    this._noHeaderRow = false
    this._dateFormat = Object.keys(DATE_FORMATS)[0]
    this._errors = { base: [], transactions: {} }
    this._currentRow = 0
  }

  get csvData() {
    return this._csvData
  }

  get csvHeader() {
    return this._csvHeader
  }

  get firstRowIndex() {
    return this._firstRowIndex
  }

  get transactionFields() {
    return TRANSACTION_FIELDS
  }

  isFieldSelected(transactionField) {
    return this._csvHeader.findIndex(column => column.transactionField === transactionField) !== -1
  }

  mapColumnToTransactionField({ columnIndex, transactionField }) {
    this._csvHeader[columnIndex].transactionField = transactionField
  }

  get dateFormats() {
    return Object.keys(DATE_FORMATS)
  }

  get dateFormat() {
    return this._dateFormat
  }

  set dateFormat(newDateFormat) {
    this._dateFormat = newDateFormat
  }

  get noHeaderRow() {
    return this._noHeaderRow
  }

  set noHeaderRow(value) {
    this._noHeaderRow = value
  }

  // Look through the first few rows and find the most common number of columns
  countColumns() {
    // crete a hash with {numColums: numRows}
    const columnCount = this._csvData.reduce((count, row) => {
      return {
        ...count,
        [row.length]: (count[row.length] || 0) + 1
      }
    }, {})

    // Select the most common column length
    const mostCommonLength = Object.keys(columnCount)
      .reduce((mostCommon, length) => (
        columnCount[length] > columnCount[mostCommon] ? length : mostCommon
      ))

    this._columnsCount = parseInt(mostCommonLength, 10)
  }

  // Find the first row that has the same number of columns as most of the other rows
  findFirstRow() {
    this._firstRowIndex = this._csvData.findIndex(row => row.length === this._columnsCount)
  }

  mapHeaderToTransactionFields() {
    let rowCount = -1
    this._csvHeader = this._csvData[this._firstRowIndex].reduce((res, columnHeader) => {
      rowCount += 1
      const transactionField = Object.keys(TRANSACTION_FIELDS).find(field => (
        TRANSACTION_FIELDS[field].regex.test(columnHeader)
      ))
      return [
        ...res,
        {
          label: columnHeader,
          transactionField: transactionField || DONT_IMPORT,
          sample: this._csvData[this._firstRowIndex + 1][rowCount]
        }
      ]
    }, [])
  }


  parse(file) {
    this._file = file
    return new Promise((resolve) => {
      Papa.parse(this._file, {
        trimHeaders: true,
        dynamicTyping: true,
        skipEmptyLines: 'greedy',
        complete: (results) => {
          this._csvData = results.data
          this.countColumns()
          this.findFirstRow()

          // Record any errors from the parsing library
          results.errors.forEach((error) => {
            this.addError(`${error.type}: (row ${error.row}) ${error.message}`, 'base')
          })

          this.mapHeaderToTransactionFields()
          this.detectDateFormat()
          return resolve()
        }
      })
    })
  }

  hasErrors() {
    return this._errors.base.length > 0 || Object.keys(this._errors.transactions).length > 0
  }

  mapToTransactions() {
    const transactions = []
    const columns = this._csvHeader.reduce((res, column, index) => {
      if (column.transactionField === DONT_IMPORT) return res
      return { ...res, [column.transactionField]: index }
    }, {})

    this._csvData.forEach((row, index) => {
      if (this._noHeaderRow || index > this._firstRowIndex) {
        const amount = this.readAmount(row, columns)
        const description = this.readDescription(row, columns)
        const createdAt = this.dateFromString(row[columns.createdAt])
        const errors = []
        if (typeof amount !== 'number') {
          errors.push('Could not read the amount')
        } else if (createdAt === null) {
          errors.push(`Invalid date. Expecting format '${this._dateFormat}'`)
        }
        transactions.push({
          amount,
          description,
          createdAt,
          errors
        })
      }
    })
    return { transactions, errors: this._errors }
  }

  readAmount(row, columns) {
    let amount
    if (Object.keys(columns).includes('amount')) {
      amount = row[columns.amount]
    } else if (Object.keys(columns).includes('income') && row[columns.income] !== null) {
      amount = row[columns.income]
    } else if (Object.keys(columns).includes('expense') && row[columns.expense] !== null) {
      amount = -row[columns.expense]
    }
    return amount
  }

  readDescription(row, columns) {
    const description = ['description1', 'description2'].reduce((acc, column) => {
      if (Object.keys(columns).includes(column)) {
        return [
          ...acc,
          (isNil(row[columns[column]]) ? '' : row[columns[column]].trim())
        ]
      }
      return acc
    }, [])
    return description.join(' - ')
  }

  dateFromString(dateString) {
    let string
    try {
      string = dateString.toString()
    } catch (error) {
      return null
    }
    try {
      const [year, month, day] = DATE_FORMATS[this._dateFormat](string)
      return parse(`${year}-${month}-${day}`).getTime()
    } catch (error) {
      return null
    }
  }

  // Adds an error either to the base which  relative to the file in general
  // or to a specific transaction
  addError(error, type) {
    if (type === 'base') {
      this._errors.base.push(error)
    } else {
      if (!(this._currentRow in this._errors.transactions)) {
        this._errors.transactions[this._currentRow] = []
      }
      this._errors.transactions[this._currentRow].push(error)
    }
  }

  // Run all the date formats by the column data and pick the one that doesn't fail
  findDateFormatFor(string) {
    return Object.keys(DATE_FORMATS).find((dateFormat) => {
      try {
        const [year, month, day] = DATE_FORMATS[dateFormat](string)
        parse(`${year}-${month}-${day}`)
      } catch (error) {
        return false
      }
      return true
    })
  }

  detectDateFormat() {
    // Do we have a date column yet?
    const createdAtIndex = this._csvHeader.findIndex(column => column.transactionField === 'createdAt')

    // We don't have a date column yet so test all the columns of each row
    if (createdAtIndex === -1) {
      const formatMap = {} // { yyyyymmdd: {columnIndex: totalMatches, ...} }}
      let dateColumnIndex = -1
      let bestDateFormat = Object.keys(DATE_FORMATS)[0]
      let totalMatches = 0
      this._csvData.some((row) => {
        row.forEach((cell, index) => {
          const dateFormat = this.findDateFormatFor(cell)
          if (dateFormat !== undefined) {
            if (!Object.keys(formatMap).includes(dateFormat)) {
              formatMap[dateFormat] = { [index]: 0 }
            }
            if (!Object.keys(formatMap[dateFormat]).includes(`${index}`)) {
              formatMap[dateFormat][index] = 0
            }
            formatMap[dateFormat][index] += 1
            // Record the best match
            if (formatMap[dateFormat][index] > totalMatches) {
              totalMatches = formatMap[dateFormat][index]
              bestDateFormat = dateFormat
              dateColumnIndex = index
            }
          }
        })
        // Leave the loop as soon as we have enought confirmed matches
        if (totalMatches >= 10) return true
        // Otherwise keep going
        return false
      })
      // Select the best match
      this._dateFormat = bestDateFormat
      // Set the Date column
      if (dateColumnIndex > -1) {
        this._csvHeader[dateColumnIndex].transactionField = 'createdAt'
      }
    } else {
      this._dateFormat = this.findDateFormatFor(this._csvHeader[createdAtIndex].sample)
    }


    // We didn't find any good format so just pick the first one
    if (this._dateFormat === null) {
      this._dateFormat = Object.keys(DATE_FORMATS)[0]
    }
  }
}
