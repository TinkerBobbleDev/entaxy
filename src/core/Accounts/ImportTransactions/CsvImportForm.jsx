/* eslint-disable no-console */
/* eslint-disable react/no-array-index-key */
import React from 'react'
import { withStyles } from '@material-ui/core/styles'
import PropTypes from 'prop-types'
import Paper from '@material-ui/core/Paper'
import Table from '@material-ui/core/Table'
import TableHead from '@material-ui/core/TableHead'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableRow from '@material-ui/core/TableRow'
import MenuItem from '@material-ui/core/MenuItem'
import FormControl from '@material-ui/core/FormControl'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import Checkbox from '@material-ui/core/Checkbox'
import Typography from '@material-ui/core/Typography'
import InputLabel from '@material-ui/core/InputLabel'
import Select from '@material-ui/core/Select'
import Button from '@material-ui/core/Button'
import Divider from '@material-ui/core/Divider'
import CircularProgress from '@material-ui/core/CircularProgress'
import green from '@material-ui/core/colors/green'
import Icon from '@mdi/react'
import { mdiFileDelimited } from '@mdi/js'
import CsvDropzone from './CsvDropzone'
import CsvParser from '../../../store/transactions/CsvParsers/CsvParser'

const styles = (theme) => {
  return ({
    root: {
      display: 'flex',
      justifyContent: 'center',
      margin: '20px'
    },
    fileDetails: {
      display: 'flex',
      padding: 5,
      background: theme.palette.secondary.light,
      color: theme.palette.primary.contrastText,
      borderRadius: 4,
      marginBottom: 10
    },
    csvFileIcon: {
      marginRight: 5,
      verticalAlign: 'text-bottom',
      fill: '#fff'
    },
    formField: {
      width: '100%'
    },
    table: {
      marginBottom: 20
    },
    tableHeader: {
      height: 20,
      background: theme.palette.grey[500]
    },
    tableCell: {
      width: 150,
      color: 'white'
    },
    sample: {
      color: theme.palette.grey[500],
      fontSize: 11,
      width: 150,
      overflow: 'hidden',
      whiteSpace: 'nowrap'
    },
    elipsys: {
      textOverflow: 'ellipsis'
    },
    formOptions: {
      display: 'flex',
      justifyContent: 'center',
      flexDirection: 'column',
      marginTop: 10,
      padding: 10,
      border: 'solid #999 2px'
    },
    formActions: {
      display: 'flex',
      'justify-content': 'flex-end',
      'padding-top': '10px'
    },
    buttonWrapper: {
      margin: theme.spacing(1),
      position: 'relative'
    },
    buttonSuccess: {
      backgroundColor: green[500],
      '&:hover': {
        backgroundColor: green[700]
      }
    },
    buttonProgress: {
      color: green[500],
      position: 'absolute',
      top: '50%',
      left: '50%',
      marginTop: -12,
      marginLeft: -12
    }
  })
}

const DONT_IMPORT = 'Don\'t import'

export class CsvImportFormComponent extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      isSubmitting: false,
      file: null,
      error: null,
      csvHeader: null,
      dateFormat: null
    }
    this.parser = new CsvParser(this.props.budgetRules)
  }

  handleFileUpload = async (acceptedFiles) => {
    if (acceptedFiles[0].type !== 'text/csv') {
      this.setState({
        file: null,
        error: 'The file you uploaded is not a CSV file'
      })
    } else {
      await this.parser.parse(acceptedFiles[0])
      this.setState({
        file: acceptedFiles[0],
        error: null,
        csvHeader: this.parser.csvHeader,
        dateFormat: this.parser.dateFormat
      })
    }
  }

  handleChange = ({ target }) => {
    this.parser.mapColumnToTransactionField({
      columnIndex: parseInt(target.name, 10),
      transactionField: target.value
    })
    this.setState({ csvHeader: this.parser.csvHeader })
  }

  handleChangeDateFormat = ({ target }) => {
    this.parser.dateFormat = target.value
    this.setState({ dateFormat: this.parser.dateFormat })
  }

  handleChangeNoHeaderRow = ({ target }) => {
    this.parser.noHeaderRow = target.checked
    this.setState({ noHeaderRow: this.parser.noHeaderRow })
  }

  handleSubmit = (event) => {
    this.setState({ isSubmitting: true })
    const { handleParsedData } = this.props
    const { transactions, errors } = this.parser.mapToTransactions()
    handleParsedData(transactions, errors)
    this.setState({ isSubmitting: false })
    event.preventDefault()
  }

  render() {
    const {
      classes,
      onCancel,
      account
    } = this.props

    const {
      file,
      error,
      isSubmitting,
      csvHeader,
      dateFormat
    } = this.state

    return (
      <form onSubmit={this.handleSubmit}>
        <div className={classes.root}>
          {(!file || error) && (
            <CsvDropzone
              className={classes.dropzone}
              handleFileUpload={this.handleFileUpload}
              account={account}
              file={file}
              error={error}
            />
          )}
          {file && !error && (
            <div>
              <Typography variant="subtitle2" className={classes.fileDetails}>
                <Icon path={mdiFileDelimited} size={0.7} className={classes.csvFileIcon} />
                {file.name}
              </Typography>
              <Typography variant="caption">
                Choose the transaction field that matches each CSV columns
              </Typography>

              <Table className={classes.table}>
                <TableHead>
                  <TableRow className={classes.tableHeader}>
                    <TableCell className={classes.tableCell} align="right">CSV column</TableCell>
                    <TableCell className={classes.tableCell}>Transaction field</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {csvHeader.map(({ label, transactionField, sample }, index) => (
                    <TableRow key={`${index}-${label}`} selected={transactionField !== DONT_IMPORT}>
                      <TableCell align="right">
                        {label}
                        <div className={classes.sample}>
                          <div className={classes.elipsys}>{sample}</div>
                        </div>
                      </TableCell>
                      <TableCell align="left">
                        <FormControl className={classes.formField}>
                          <Select
                            value={transactionField}
                            onChange={this.handleChange}
                            inputProps={{ name: `${index}` }}
                          >
                            <MenuItem key={DONT_IMPORT} value={DONT_IMPORT}>
                              <Typography color="textSecondary">{DONT_IMPORT}</Typography>
                            </MenuItem>
                            {Object.keys(this.parser.transactionFields).map(field => (
                              <MenuItem
                                key={field}
                                value={field}
                                disabled={this.parser.isFieldSelected(field)}
                              >
                                {this.parser.transactionFields[field].label}
                              </MenuItem>
                            ))}
                          </Select>
                        </FormControl>
                      </TableCell>
                    </TableRow>
                  ))}
                </TableBody>
              </Table>
              <Typography variant="subtitle2">Import options</Typography>
              <Paper elevation={0} className={classes.formOptions}>
                <FormControlLabel
                  className={classes.formField}
                  control={(
                    <Checkbox
                      checked={this.state.noHeaderRow}
                      onChange={this.handleChangeNoHeaderRow}
                      value="noHeaderRow"
                    />
                  )}
                  label="No header row"
                />
                <FormControl className={classes.formField}>
                  <InputLabel htmlFor="dateFormat">Date format</InputLabel>
                  <Select
                    value={dateFormat}
                    onChange={this.handleChangeDateFormat}
                    inputProps={{ name: 'dateFormat' }}
                  >
                    {this.parser.dateFormats.map(format => (
                      <MenuItem key={format} value={format}>{format}</MenuItem>
                    ))}
                  </Select>
                </FormControl>
              </Paper>
            </div>
          )}
        </div>
        <Divider />
        <div className={classes.formActions}>
          <div className={classes.buttonWrapper}>
            <Button
              type="submit"
              variant="contained"
              color="primary"
              disabled={isSubmitting || file === null || error !== null}
              onClick={this.handleButtonClick}
            >
              Import
            </Button>
            {isSubmitting && <CircularProgress size={24} className={classes.buttonProgress} />}
          </div>
          <Button onClick={onCancel} color="secondary">Cancel</Button>
        </div>
      </form>
    )
  }
}

CsvImportFormComponent.propTypes = {
  classes: PropTypes.object.isRequired,
  handleParsedData: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  account: PropTypes.object.isRequired,
  budgetRules: PropTypes.object.isRequired
}

export default withStyles(styles)(CsvImportFormComponent)
