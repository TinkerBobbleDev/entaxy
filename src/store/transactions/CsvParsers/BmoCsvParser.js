import CsvParser from './CsvParser'

export default class BmoCsvParser extends CsvParser {
  constructor() {
    super()
    this._config = {
      ...this._config,
      comments: 'Following data is valid as of' // Skipp these lines
    }
    this._header = [
      'First Bank Card',
      'Transaction Type',
      'Date Posted',
      'Transaction Amount',
      'Description'
    ]
  }

  map(row, accountData) {
    return {
      ...accountData,
      type: (row['Transaction Amount'] >= 0 ? 'buy' : 'sell'),
      shares: row['Transaction Amount'],
      bookValue: 1,
      description: this.parseString(row.Description),
      createdAt: this.parseDate(row['Date Posted'], 'yyyymmdd')
    }
  }
}
