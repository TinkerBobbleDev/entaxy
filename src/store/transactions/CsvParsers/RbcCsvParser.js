import uuid from 'uuid/v4'
import CsvParser from './CsvParser'

export default class RbcCsvParser extends CsvParser {
  constructor() {
    super()
    this._header = [
      'Account Type',
      'Account Number',
      'Transaction Date',
      'Cheque Number',
      'Description 1',
      'Description 2',
      'CAD$',
      'USD$'
    ]
  }

  map(row) {
    const amount = (row.CAD$ === null ? parseFloat(row.USD$) : parseFloat(row.CAD$))
    return {
      id: uuid(),
      institution: 'RBC',
      account: row['Account Type'],
      type: (amount >= 0 ? 'buy' : 'sell'),
      ticker: (row.CAD$ === null ? 'USD' : 'CAD'),
      shares: amount,
      bookValue: 1,
      description: this.parseString(`${row['Description 1']} - ${row['Description 1']}`),
      createdAt: this.parseDate(row['Transaction Date'], 'mm/dd/yyyy')
    }
  }
}
