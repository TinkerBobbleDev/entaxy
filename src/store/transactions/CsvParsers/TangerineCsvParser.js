import CsvParser from './CsvParser'

export default class TangerineCsvParser extends CsvParser {
  constructor() {
    super()
    this._header = [
      'Date',
      'Transaction',
      'Name',
      'Memo',
      'Amount'
    ]
  }

  map(row, accountData) {
    return {
      ...accountData,
      amount: row.Amount,
      description: this.parseString(row.Memo),
      createdAt: this.parseDate(row.Date, 'mm/dd/yyyy')
    }
  }
}
