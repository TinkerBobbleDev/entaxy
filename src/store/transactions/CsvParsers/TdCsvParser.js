import CsvParser from './CsvParser'

export default class BmoCsvParser extends CsvParser {
  constructor() {
    super()
    this._config = {
      ...this._config,
      header: false // This CSV file has no header
    }
  }

  // [0] Date
  // [1] Description
  // [2] Withdrawals
  // [3] Deposits
  // [4] Balance
  map(row, accountData) {
    return {
      ...accountData,
      type: (row[2] === null ? 'buy' : 'sell'),
      shares: (row[2] === null ? row[3] : row[2]),
      bookValue: 1,
      description: this.parseString(row[1]), // Description
      createdAt: this.parseDate(row[0], 'mm/dd/yyyy')
    }
  }
}
