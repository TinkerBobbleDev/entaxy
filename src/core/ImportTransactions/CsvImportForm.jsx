import React from 'react'
import { compose } from 'recompose'
import { withStyles } from '@material-ui/core/styles'
import PropTypes from 'prop-types'
import { withFormik } from 'formik'
import Button from '@material-ui/core/Button'
import Divider from '@material-ui/core/Divider'
import CircularProgress from '@material-ui/core/CircularProgress'
import green from '@material-ui/core/colors/green'
import CsvDropzone from './CsvDropzone'
import CsvImportFields from './CsvImportFields'
import RbcCsvParser from '../../store/transactions/CsvParsers/RbcCsvParser'
import BmoCsvParser from '../../store/transactions/CsvParsers/BmoCsvParser'
import TdCsvParser from '../../store/transactions/CsvParsers/TdCsvParser'
import TangerineCsvParser from '../../store/transactions/CsvParsers/TangerineCsvParser'

const styles = theme => ({
  root: {
    display: 'flex',
    'justify-content': 'center'
  },
  input: {
    margin: '5px',
    width: 200
  },
  formActions: {
    display: 'flex',
    'justify-content': 'flex-end',
    'padding-top': '10px'
  },
  wrapper: {
    margin: theme.spacing.unit,
    position: 'relative'
  },
  buttonSuccess: {
    backgroundColor: green[500],
    '&:hover': {
      backgroundColor: green[700]
    }
  },
  buttonProgress: {
    color: green[500],
    position: 'absolute',
    top: '50%',
    left: '50%',
    marginTop: -12,
    marginLeft: -12
  }
})

const parsers = {
  RBC: RbcCsvParser,
  BMO: BmoCsvParser,
  TD: TdCsvParser,
  Tangerine: TangerineCsvParser
}

class CsvImportForm extends React.Component {
  state = {
    file: null,
    error: null
  }

  handleFileUpload = (acceptedFiles) => {
    this.setState({
      file: acceptedFiles[0],
      error: (acceptedFiles[0].type === 'text/csv' ? null : 'The file you uploaded is not a CSV file')
    })
    this.props.setFieldValue('file', acceptedFiles[0])
  }

  render() {
    const {
      classes,
      handleSubmit,
      values,
      handleChange,
      onCancel,
      institution,
      isSubmitting
    } = this.props

    const {
      file,
      error
    } = this.state

    return (
      <form onSubmit={handleSubmit}>
        <div className={classes.root}>
          <CsvImportFields
            institution={institution}
            values={values}
            handleChange={handleChange}
          />
          <CsvDropzone
            className={classes.dropzone}
            handleFileUpload={this.handleFileUpload}
            institution={institution}
            file={file}
            error={error}
          />
        </div>
        <Divider />
        <div className={classes.formActions}>
          <div className={classes.wrapper}>
            <Button
              type="submit"
              variant="contained"
              color="primary"
              disabled={isSubmitting || file === null || error !== null}
              onClick={this.handleButtonClick}
            >
              Import
            </Button>
            {isSubmitting && <CircularProgress size={24} className={classes.buttonProgress} />}
          </div>
          <Button onClick={onCancel} color="secondary">Cancel</Button>
        </div>
      </form>
    )
  }
}

CsvImportForm.propTypes = {
  classes: PropTypes.object.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  values: PropTypes.object.isRequired,
  handleChange: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  institution: PropTypes.string.isRequired,
  setFieldValue: PropTypes.func.isRequired,
  isSubmitting: PropTypes.bool.isRequired
}

export default compose(
  withStyles(styles),
  withFormik({
    mapPropsToValues: ({ institution }) => ({
      institution,
      account: '',
      type: '',
      ticker: '',
      shares: '',
      bookValue: '',
      createdAt: new Date(),
      file: null
    }),
    handleSubmit: (values, { props, setSubmitting }) => {
      setSubmitting(true)
      new parsers[values.institution]().parse(values.file, values)
        .then(({ transactions, errors }) => {
          setSubmitting(false)
          props.handleParsedData(transactions, errors)
        })
    }
  })
)(CsvImportForm)
