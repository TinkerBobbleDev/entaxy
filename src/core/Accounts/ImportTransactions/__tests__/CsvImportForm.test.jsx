import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import { CsvImportFormComponent } from '../CsvImportForm'
import { DATE_FORMATS } from '../../../../store/transactions/CsvParsers/CsvParser'


// jest.mock('../../../../store/transactions/CsvParsers/CsvParser')
// const CsvParser = require('../../../../store/transactions/CsvParsers/CsvParser')

describe('CsvImportForm', () => {
  const mockHandleParsedData = jest.fn()
  const mockOnCancel = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('matches snapshot for BMO', () => {
    const component = renderer.create((
      <CsvImportFormComponent
        handleParsedData={mockHandleParsedData}
        onCancel={mockOnCancel}
        account={{ id: 1, institution: 'BMO' }}
        transactionRules={{ }}
        classes={{ }}
      />
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })

  describe('Component methods', () => {
    let instance
    let parseSpy
    let mapToTransactionsSpy
    let csvData

    beforeEach(() => {
      const wrapper = shallow((
        <CsvImportFormComponent
          handleParsedData={mockHandleParsedData}
          onCancel={mockOnCancel}
          account={{ id: 1, institution: 'BMO' }}
          transactionRules={{ }}
          classes={{ }}
        />
      ))
      instance = wrapper.instance()
      parseSpy = jest.spyOn(instance.parser, 'parse')
      mapToTransactionsSpy = jest.spyOn(instance.parser, 'mapToTransactions')
      csvData = [
        'First Bank Card,Transaction Type,Date Posted, Transaction Amount,Description',
        '\'500766**********\',DEBIT,20180628,-650.0,[SO]2211#8503-567 ',
        '\'500766**********\',CREDIT,20180629,2595.11,[DN]THE WORKING GRO PAY/PAY  '
      ]
    })

    it('should handleFileUpload', async () => {
      const file = new File([csvData.join('\n')], 'test.csv', { type: 'text/csv' })
      await instance.handleFileUpload([file])
      expect(parseSpy).toHaveBeenCalledWith(file)
      expect(instance.state).toEqual({
        file,
        error: null,
        csvHeader: instance.parser.csvHeader,
        dateFormat: instance.parser.dateFormat,
        isSubmitting: false
      })
      expect(instance.state.csvHeader.map(item => (item.label)).join()).toEqual(csvData[0])
    })

    it('should handleFileUpload for invalid file', async () => {
      const file = new File([csvData.join('\n')], 'test.csv', { type: 'text/xls' })
      await instance.handleFileUpload([file])
      expect(parseSpy).not.toHaveBeenCalled()
      expect(instance.state).toEqual({
        file: null,
        error: 'The file you uploaded is not a CSV file',
        csvHeader: null,
        dateFormat: null,
        isSubmitting: false
      })
    })

    it('should handleChange', async () => {
      const file = new File([csvData.join('\n')], 'test.csv', { type: 'text/csv' })
      await instance.handleFileUpload([file])
      expect(instance.state.csvHeader[3]).toEqual({
        label: ' Transaction Amount',
        sample: -650,
        transactionField: 'amount'
      })
      instance.handleChange({ target: { value: 'Don\'t import', name: 3 } })
      expect(instance.state.csvHeader[3]).toEqual({
        label: ' Transaction Amount',
        sample: -650,
        transactionField: 'Don\'t import'
      })
      instance.handleChange({ target: { value: 'amount', name: 3 } })
      expect(instance.state.csvHeader[3]).toEqual({
        label: ' Transaction Amount',
        sample: -650,
        transactionField: 'amount'
      })
    })

    it('should handleChangeDateFormat', async () => {
      const file = new File([csvData.join('\n')], 'test.csv', { type: 'text/csv' })
      await instance.handleFileUpload([file])
      expect(instance.parser.dateFormat).toEqual(Object.keys(DATE_FORMATS)[1])
      instance.handleChangeDateFormat({ target: { value: Object.keys(DATE_FORMATS)[0] } })
      expect(instance.parser.dateFormat).toEqual(Object.keys(DATE_FORMATS)[0])
      expect(instance.state.dateFormat).toEqual(Object.keys(DATE_FORMATS)[0])
    })

    it('should handleChangeNoHeaderRow', async () => {
      const file = new File([csvData.join('\n')], 'test.csv', { type: 'text/csv' })
      await instance.handleFileUpload([file])
      expect(instance.parser.noHeaderRow).toBeFalsy()
      instance.handleChangeNoHeaderRow({ target: { checked: true } })
      expect(instance.parser.noHeaderRow).toBeTruthy()
      expect(instance.state.noHeaderRow).toBeTruthy()
    })

    it('should handleSubmit', async () => {
      const mockEvent = { preventDefault: jest.fn() }

      const file = new File([csvData.join('\n')], 'test.csv', { type: 'text/csv' })
      await instance.handleFileUpload([file])
      instance.handleSubmit(mockEvent)
      expect(mapToTransactionsSpy).toHaveBeenCalled()
      expect(mockHandleParsedData).toHaveBeenCalled()
      expect(mockEvent.preventDefault).toHaveBeenCalled()
    })
  })
})
