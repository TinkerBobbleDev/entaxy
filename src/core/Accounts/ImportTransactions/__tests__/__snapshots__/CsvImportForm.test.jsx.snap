// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CsvImportForm matches snapshot for BMO 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <div
      className="CsvDropzone-root-1"
      onBlur={[Function]}
      onClick={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      <input
        autoComplete="off"
        multiple={false}
        onChange={[Function]}
        onClick={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        tabIndex={-1}
        type="file"
      />
      <div>
        <h4
          className="MuiTypography-root MuiTypography-h4 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
        >
          Drag a CSV file here
        </h4>
        <h6
          className="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
        >
          or click to select a file to upload.
        </h6>
      </div>
    </div>
  </div>
  <hr
    className="MuiDivider-root"
  />
  <div>
    <div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
        disabled={true}
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={-1}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          Import
        </span>
      </button>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary"
      disabled={false}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Cancel
      </span>
    </button>
  </div>
</form>
`;
